src/System.Device.Gpio/System/Device/Gpio/Drivers/LibGpiodDriver.Linux.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Runtime.InteropServices;
+using System.IO;
+
+namespace System.Device.Gpio.Drivers
+{
+    public class LibGpiodDriver : UnixDriver
+    {
+        private SafeChipHandle _chip;
+
+        private Dictionary<int, SafeLineHandle> _pinNumberToSafeLineHandle;
+
+        protected internal override int PinCount => Interop.GetNumberOfLines(_chip);
+
+        public LibGpiodDriver()
+        {
+            SafeChipIteratorHandle iterator = Interop.GetChipIterator();
+            if (iterator == null)
+            {
+                throw new IOException($"No any chip available, error code: {Marshal.GetLastWin32Error()}");
+            }
+
+            _chip = Interop.GetNextChipFromChipIterator(iterator);
+            if (_chip == null)
+            {
+                throw new IOException($"No chip found, error code: {Marshal.GetLastWin32Error()}");
```

Comment:
Consider using a throw helper pattern to reduce the method size bloat that might come from formatting the error message in line.

You can either call a helper which returns the exception you want to throw (which contains the exception message), or defer throwing all together to the helper.

Example [sharplab.io](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxNMAfAAgBgARYCMA3ALABQWAzAQEx4DCeA3hXuwTcAPbcA2eALKEAFEXwA3CHwCucJHgCWAOxh4+cZQHMYACwCULNhxOKAZnhFTZcPAF47eZTL599xk+1blPvvHoRuAHcnOBCAUQAPAGM4AAcYRW5lEQAiPAA5FwFrOTxU/TIfPwBfD08sAHZ/BDkikzLi9nLqPB5+IVoxQklpOQUVNQ1tPUNvPyULKz7bByds9ybPcYmOAODQiJj4xOSRABJ0rNc8XNtmYZ1dEoL60vKTKpq68saTFq5eAUEqbt6bAaqdSaK5jB4ccyWM72RzOVyLCYrVZ4AAqukCQQAkgB5KKxBJJFKXUbEPDg9hvPxPGC1OB3CnlcpxBCKKTwAiEABsBAALKj0cEcXidoSRINgSMDOUkZ51iFlGE8MKCXtDplsqcZixiddbq8KCUgA=)

Reply:
@AndyAyersMS, what are your thoughts on using the format string pattern to throw exception messages within the method which contains the happy path?

Reply:
It can potentially add overhead to the method prolog and epilog, so if that overhead impacts perf of the happy path then you might consider using a throw helper.

Reply:
@ahsonkhan looked into the sample you referenced, seems good approach when error messages use similar string lateral, but here each errornous case messages are different. 

Also not sure about what you mean by happy path, if you mean about normal program execution path then: no it is not happy path, program execution will stop if any of these exceptions thrown

Reply:
> but here each errornous case messages are different
In corefx, we have used an exception enum to get the appropriate string message within the throw helper.

Reply:
So something like this:
```C#
private static class ThrowHelper
{
   // Add overloads for other exception types or exceptions with other format string data as parameters
   public static void ThrowIOException(ExceptionResource resource, int errorCode /*, ... */)
   {
      throw GetIOException(GetResourceString(resource, errorCode));
   }
   private static IOException GetIOException(string message)
   {
      return new IOException(message);
   }
   private static string GetResourceString(ExceptionResource resource, int errorCode /*, ... */)
   {
      string message = "";
      switch(resource)
      {
         case ExceptionResource.NoChipFound:
            message = $"No chip found, error code: {errorCode}";
            break;
         // ... More cases
      }
      return message;
   }
}
internal enum ExceptionResource
{
   NoChipFound,
}
```

Usage, in this case, would become something like:
```C#
   _chip = Interop.GetNextChipFromChipIterator(iterator);
   if (_chip == null)
   {
      ThrowHelper.ThrowIOException(ExceptionResource.NoChipFound, Marshal.GetLastWin32Error());
```

Reply:
Looks good, will update, thanks